File: Problems/0400-nth-digit.py
class Solution(object):
    def findNthDigit(self, n):
        """
        :type n: int
        :rtype: int
        """
        digits = 1
        start = 1
        count = 9

        while n > digits * count:
            n -= digits * count
            digits += 1
            start *= 10
            count *= 10

        num = start + (n - 1) // digits
        return int(str(num)[(n - 1) % digits])

File: Problems/0400-nth-digit.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here
class Solution {
public:
    int findNthDigit(int n) {
        long long digits = 1;
        long long start = 1;
        long long count = 9;

        while (n > digits * count) {
            n -= digits * count;
            digits++;
            start *= 10;
            count *= 10;
        }

        long long num = start + (n - 1) / digits;
        string s = to_string(num);

        return s[(n - 1) % digits] - '0';
    }
};